hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala

```diff
@@ -136,27 +126,29 @@ object ArrayFunctions extends RegistryFunctions {
 
     registerIR("product", TArray(tnum("T")), tv("T"))(product)
 
-    def makeMinMaxOp(op: String): IR => IR = {
-      { a =>
+    def makeMinMaxOp(op: String): Seq[IR] => IR = {
+      { case Seq(a) =>
         val t = -coerce[TArray](a.typ).elementType
-        val accum = genUID()
         val value = genUID()
-
-        val aUID = genUID()
-        val aRef = Ref(aUID, a.typ)
-        val zVal = If(ApplyComparisonOp(EQ(TInt32()), ArrayLen(aRef), I32(0)), NA(t), ArrayRef(a, I32(0)))
-
-        val body = invoke(op, t, Ref(value, t), Ref(accum, t))
-        Let(aUID, a, ArrayFold(aRef, zVal, accum, value, body))
+        val first = genUID()
+        val acc = genUID()
+        ArrayFold2(a,
+          FastIndexedSeq((acc, NA(t)), (first, True())),
+          value,
+          FastIndexedSeq(
+            If(Ref(first, TBoolean()), Ref(value, t), invoke(op, t, Ref(acc, t), Ref(value, t))),
+            False()
+          ),
+          Ref(acc, t))
```

Comment:
Keeping the boolean state seems unnecessary. What about only doing a fold when the array is non-empty, like:
```scala
val t = -coerce[TArray](a.typ).elementType
val accum = genUID()
val value = genUID()
val aUID = genUID()
val aRef = Ref(aUID, a.typ)
val zVal = ArrayRef(a, I32(0))
val body = invoke(op, t, Ref(value, t), Ref(accum, t))

Let(aUID, a,
  If(ApplyComparisonOp(EQ(TInt32()), ArrayLen(aRef), I32(0)),
    NA(t),
    ArrayFold(aRef, zVal, accum, value, body)))
```

Reply:
(as discussed -- wrapping in a Let prevents deforestation)